src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -3613,23 +3621,21 @@
 
             if (ignoreCase)
             {
-                // This linear search is not the optimal way to use a hashmap, but we should only use
-                // this functionality when computing diagnostics.  Note
-                // that we can't store the map case-insensitively, since real metadata name
-                // lookup has to remain case sensitive.
-                foreach (var pair in _lazyForwardedTypesToAssemblyIndexMap)
+                // We should only use this functionality when computing diagnostics, so we lazily construct
+                // a case-insensitive map when necessary. Note that we can't store the original map
+                // case-insensitively, since real metadata name lookup has to remain case sensitive.
+                ensureCaseInsensitiveDictionary();
+
+                if (_lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap.TryGetValue(fullName, out var value))
                 {
-                    if (string.Equals(pair.Key, fullName, StringComparison.OrdinalIgnoreCase))
-                    {
-                        matchedName = pair.Key;
-                        return pair.Value;
-                    }
+                    var found = _lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap.TryGetKey(fullName, out matchedName);
```

Comment:
>out matchedName [](http://example.com/codeflow?start=105&length=15)

We probably should add a test to cover this aspect (return of the actual spelling)

Reply:
Will try and devise a test that hits this codepath.

Reply:
It looks like it is not possible to observe this. https://github.com/dotnet/roslyn/blob/main/src/Compilers/VisualBasic/Portable/Binding/Binder_Symbols.vb#L421 is the only codepath that calls this function with `ignoreCase=True`, and it does not use the `ByRef metadataName` parameter again after the difference would have been noticed.

Reply:
There is an existing test I would have expected to observe the difference, if any difference were able to be observed: https://github.com/dotnet/roslyn/blob/main/src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/TypeForwarders.vb#L745-L782